src/dmd/typesem.d

```diff
@@ -3080,9 +3081,30 @@ private extern(C++) final class DotExpVisitor : Visitor
                 /* Rewrite e.ident as:
                  *  e.aliasthis.ident
                  */
-                e = resolveAliasThis(sc, e);
-                auto die = new DotIdExp(e.loc, e, ident);
-                return returnExp(die.semanticY(sc, flag & 1));
+                // gag always: we don't know if we'll actually
+                // succeed on the DotIdExp access later on, and if we fail it, we
+                // want to suppress warnings arising from the alias this access.
+                auto alias_e = resolveAliasThis(sc, e, true);
+
+                if (!alias_e)
+                {
+                    if (!gagError)
+                        resolveAliasThis(sc, e, false); // but we do want to report errors.
```

Comment:
Why is this necessary? Won't the errors be reported at line 3104?

Reply:
It it's necessary, why is `if (!gagError)` necessary?

Reply:
No, because we're bailing here - we can't continue because alias_e is null, so we can't make a DotIdExp of it. This can happen if the target of the alias is itself invalid. This is the part that fixes Issue 12807.

Reply:
I guess I could set e to the ungagged expression because it'll know that it's erroring.

Reply:
No I can't! Because even so, the later property access could still fail a la 18775. So I cannot error that early ever, anyways.

I stand by the current impl.

Reply:
... No, we can keep going if we're not gagged. Because we'll be gagged via flag for the test to decide whether we can do ufcs after? ....? ???

Reply:
Oh, fuck. Strictly speaking I can only emit deprecations if they are *not* errors.

Reply:
Because if they *are* errors, we may want to fall through into ufcs?

Reply:
... I'll stick my neck out and say that the code as written _possibly_ does the Right Thing.

I think it definitely does a significantly righter thing than DMD does at present.

Reply:
Hm. Maybe I've been coming at this wrong.

Reply:
Okay! Here's the __issue__.

The problem is **warnings**. If I get an error in the alias resolve, I can just keep going with the errored exp; if it returns null, I can just return null since we'd be gagged. The problem is if the initial alias resolve prints a **warning**, then later on the dot lookup **fails** and we fall through to UFCS, we'll have printed a warning spuriously. That's the case that I have to workaround with the repeated resolveAliasThis calls.

Reply:
I actually managed to find an improvement.

This code is fucking *hellish*.